cmd/istio-cni-daemonset/main.go

```diff
@@ -0,0 +1,240 @@
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+	client "k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"istio.io/cni/pkg/repair"
+	"istio.io/pkg/log"
+)
+
+type ControllerOptions struct {
+	DaemonPollPeriod int             `json:"daemon_poll_period"`
+	RepairOptions    *repair.Options `json:"repair_options"`
+	DeletePods       bool            `json:"delete_pods"`
+	LabelPods        bool            `json:"label_pods"`
+	CreateEvents     bool            `json:"create_events"`
+	RunAsDaemon      bool            `json:"run_as_daemon"`
+}
+
+var (
+	loggingOptions = log.DefaultOptions()
+)
+
+// Parse command line options
+func parseFlags() (filters *repair.Filters, options *ControllerOptions) {
+	// Parse command line flags
+	// Filter Options
+	pflag.String("node-name", "", "The name of the node we are managing (will manage all nodes if unset)")
+	pflag.String(
+		"sidecar-annotation",
+		"sidecar.istio.io/status",
+		"An annotation key that indicates this pod contains an istio sidecar. All pods without this annotation will be ignored."+
+			"The value of the annotation is ignored.")
+	pflag.String(
+		"init-container-name",
+		"istio-init",
+		"The name of the istio init container (will crash-loop if CNI is not configured for the pod)")
+	pflag.String(
+		"init-container-termination-message",
+		"",
+		"The expected termination message for the init container when crash-looping because of CNI misconfiguration")
+	pflag.Int(
+		"init-container-exit-code",
+		126,
+		"Expected exit code for the init container when crash-looping because of CNI misconfiguration")
+
+	pflag.StringSlice("label-selectors", []string{}, "A set of label selectors in label=value format that will be added to the pod list filters")
+	pflag.StringSlice("field-selectors", []string{}, "A set of field selectors in label=value format that will be added to the pod list filters")
+
+	// Repair Options
+	pflag.Bool("delete-pods", false, "Controller will delete pods")
+	pflag.Bool("label-pods", false, "Controller will label pods")
+	pflag.Bool("create-events", true, "Controller will create Kubernetes events")
+	pflag.Bool("run-as-daemon", false, "Controller will run in a loop")
+	pflag.Int("daemon-poll-period", 10, "Polling period for daemon (in seconds)")
+	pflag.String(
+		"broken-pod-label-key",
+		"cni.istio.io/uninitialized",
+		"The key portion of the label which will be set by the reconciler if --label-pods is true")
+	pflag.String(
+		"broken-pod-label-value",
+		"true",
+		"The value portion of the label which will be set by the reconciler if --label-pods is true")
+
+	pflag.Bool("help", false, "Print usage information")
+
+	pflag.Parse()
+	if err := viper.BindPFlags(pflag.CommandLine); err != nil {
+		log.Fatal("Error parsing command line args: %+v")
+	}
+
+	if viper.GetBool("help") {
+		pflag.Usage()
+		os.Exit(0)
+	}
+
+	viper.SetEnvPrefix("icnid")
```

Comment:
what is icdni?

Reply:
Shortening of `istio-cni-daemonset` so that you can set ENV variables instead of passing command line arguments, e.g. '--node-name=Test' could be set by using the env variable 'ICNID_NODE-NAME=Test'

Reply:
>We already have a CNI daemonset, are users expected to run a second daemonset as well?

I could integrate this with the existing daemonset, by adding it to the install-cni.sh script.

Reply:
ICDNI changed to REPAIR, and repair code merged into existing daemonset as a new container.